rclcpp_action/include/rclcpp_action/server.hpp

```diff
@@ -356,16 +356,20 @@ class Server : public ServerBase, public std::enable_shared_from_this<Server<Act
   CancelResponse
   call_handle_cancel_callback(const GoalUUID & uuid) override
   {
-    std::lock_guard<std::mutex> lock(goal_handles_mutex_);
+    std::shared_ptr<ServerGoalHandle<ActionT>> goal_handle;
+    {
+      std::lock_guard<std::mutex> lock(goal_handles_mutex_);
+      auto element = goal_handles_.find(uuid);
+      if (element != goal_handles_.end()) {
+        goal_handle = element->second.lock();
+      }
+    }
+
     CancelResponse resp = CancelResponse::REJECT;
-    auto element = goal_handles_.find(uuid);
-    if (element != goal_handles_.end()) {
-      std::shared_ptr<ServerGoalHandle<ActionT>> goal_handle = element->second.lock();
-      if (goal_handle) {
-        resp = handle_cancel_(goal_handle);
-        if (CancelResponse::ACCEPT == resp) {
-          goal_handle->_cancel_goal();
-        }
+    if (goal_handle) {
+      resp = handle_cancel_(goal_handle);
+      if (CancelResponse::ACCEPT == resp) {
+        goal_handle->_cancel_goal();
```

Comment:
I was concerned that this introduced a race here, allowing the goal to be finished asynchronously, after it was found but before we call the user callback on this line. But it seems that was always the expectation, since the user callback can return `CancelResponse::ACCEPT` or `CancelResponse::REJECT` (though that might only be one reason for it.

This is why you need the new code in the test:

```c++
        if (!handle->is_active()) {
          return rclcpp_action::CancelResponse::REJECT;
        }
```

However, now I'm wondering about `goal_handle->_cancel_goal();`, what if this happens:

- user callback checks `if (!handle->is_active())` which fails
- user returns `CancelResponse::ACCEPT`
- (⚡) context switch, goal finishes (terminates? enters terminal state? moves to aborted/succeeded state? not sure on the terminology)
- switch back to original thread, call `goal_handle->_cancel_goal()` because `CancelResponse::ACCEPT == resp`
- now we're canceling a finished goal

It looks like `_cancel_goal` doesn't care if the state changed under us:

https://github.com/ros2/rclcpp/blob/1037822a63330495dcf0de5e8f20544375a5f116/rclcpp_action/src/server_goal_handle.cpp#L80-L88

Though `rcl` does:

https://github.com/ros2/rcl/blob/414fdbd479b83f2b8af8235ad1fa3aeba2b0027f/rcl_action/src/rcl_action/goal_handle.c#L92-L99

Maybe we should be calling `rcl_action_goal_handle_is_cancelable()` before assigning it, or catch that rcl error?

Also, it should be mentioned to users (maybe already documented) that even though their cancel goal callback is called, and they return `CancelResponse::ACCEPT`, the resulting state may not be `canceled`...

@jacobperron any insight? I'm unfortunately not as versed in the design/implementation of actions as I think I need to be to judge this.

See also: https://design.ros2.org/articles/actions.html

Reply:
@wjwwood Thanks for looking into it.

Yes, this patch could only fix the deadlock issue.

> However, now I'm wondering about goal_handle->_cancel_goal();, what if this happens:
user callback checks if (!handle->is_active()) which fails
user returns CancelResponse::ACCEPT
(⚡) context switch, goal finishes (terminates? enters terminal state? moves to aborted/succeeded state? not sure on the terminology)
switch back to original thread, call goal_handle->_cancel_goal() because CancelResponse::ACCEPT == resp
now we're canceling a finished goal

Actually this case I had mentioned in https://github.com/ros2/rclcpp/issues/1599#issuecomment-814145969, but maybe I didn't explain it clearly enough .

> Maybe we should be calling rcl_action_goal_handle_is_cancelable() before assigning it, or catch that rcl error?

> Also, it should be mentioned to users (maybe already documented) that even though their cancel goal callback is called, and they return CancelResponse::ACCEPT, the resulting state may not be canceled...

I think catching the exception internally is a straightforward way to resolve it, then the users have not to add `if (!handle->is_active())` to there CancelCallback.

But @fujitatomoya thought it was a strange behavior. https://github.com/ros2/rclcpp/issues/1599#issuecomment-810283121. So far I have no other solotions to fix this issue.

Reply:
removing mutex before calling user callback looks okay to me 👍 

> I think catching the exception internally is a straightforward way to resolve it

i believe the point for this discussion is how it appears to be for user application. currently user application does receive exception if this happens. which i think it seems to be okay since user application gets this corner case as an exception. (probably exception message and adding check `rcl_action_goal_handle_is_cancelable` would be better.) i mean basically it does what needs to be done. 

i just do not have good feeling that system conceals this exception from user application, which is suggested on https://github.com/ros2/rclcpp/issues/1599.

Reply:
I see.

> even though their cancel goal callback is called, and they return CancelResponse::ACCEPT, the resulting state may not be canceled..

In addition, I wonder make sure if you agree with this solution?

And how we fix it? catching the exception like below code? or doing some changes like this patch https://github.com/KavenYau/rclcpp/commit/3ba7b2ef01449929b6c976bdb9f21c886dff3464? or other better suggestions?
```c++
try {
  goal_handle->_cancel_goal();
} catch (const rclcpp::exceptions::RCLError & ex) {
  return CancelResponse::REJECT;
}
```

Reply:
> Maybe we should be calling rcl_action_goal_handle_is_cancelable() before assigning it, or catch that rcl error?

It looks like an issue unrelated to this change. I think that it would be nice to handle this race between transitioning to canceling and some terminal state (succeeded or aborted) instead of throwing an exception.

> Also, it should be mentioned to users (maybe already documented) that even though their cancel goal callback is called, and they return CancelResponse::ACCEPT, the resulting state may not be canceled...

Indeed, by design *accepting* a cancel request only tries to transition the goal to the **canceling** state; it does not mean the goal is immediately transitioned to the **canceled** state. I don't know if this is documented in rclcpp_action very well, but I think the design doc captures it?

Reply:
> It looks like an issue unrelated to this change.

i think so too, this fix is unrelated.

> However, it's probably better to address this in a separate PR.

agree, issue is already opened. https://github.com/ros2/rclcpp/issues/1599

> Indeed, by design accepting a cancel request only tries to transition the goal to the canceling state; it does not mean the goal is immediately transitioned to the canceled state. I don't know if this is documented in rclcpp_action very well, but I think the design doc captures it?

i did check the action design, and i see `The list of goals in the response indicates which goals will be attempted to be canceled.` as you mentioned. besides, 

![](https://design.ros2.org/img/actions/goal_state_machine.png)

it described clearly that after accepting `cancel request` that still possible transit into `succeeded`.

so i think my assumption is wrong, it should not throw exception in this case.

Reply:
I think we can follow up on this in https://github.com/ros2/rclcpp/issues/1599